### PR TITLE
Port FSTFieldPath and FSTResourcePath to C++

### PR DESCRIPTION
**Pull request by var-const**
FIRFieldPath is next.

EDIT: since, as far as I can see, nothing in the port relies on FIRFieldPath, I'm going to reverse that and actually leave off FIRFieldPath until the very end.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/model/base_path.h [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+/**
+ * BasePath represents a path sequence in the Firestore database. It is composed
+ * of an ordered sequence of string segments.
+ *
+ * BasePath is immutable. All mutating operations return new independent
+ * instances.
+ *
+ * ## Subclassing Notes
+ *
+ * BasePath is strictly meant as a base class for concrete implementations. It
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/base_path.h]: I can think of several ways to implement this:
- FieldPath inherits from BasePath, using CRTP so that methods like Concat can return the correct type. This corresponds well to the current implementation in Objective C and Java. One downside is that the interface is split among classes/files. The other downside is that it's possible to mistakenly use BasePath in a polymorphic context (I tried making it less likely by protecting its constructors/destructor and placing it inside an "impl" namespace);
- FieldPath contains BasePath, which is completely hidden from other code. Avoids all problems from the previous bullet, but requires writing a lot of boilerplate to forward all methods;
- implement all methods as template free functions. This will probably feel awkward to use.

@wilhuff Gil, I remember you mentioned you see the concrete classes as templates. What did you have in mind?

File: Firestore/core/src/firebase/firestore/model/base_path.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+class BasePath {
+ protected:
+  using SegmentsT = std::vector<std::string>;
+
+ public:
+  using const_iterator = SegmentsT::const_iterator;
+
+  /** Returns i-th segment of the path. */
+  const std::string& operator[](const size_t i) const {
+    return at(i);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/base_path.h]: Throughout the PR, I intentionally deviated from Objective C naming/etc. Feel free to push back on this.

In this case, I tried giving more idiomatic C++ names to the methods. By the way, why do we have both index operator and `at` (differently named) in Objective-C?

File: Firestore/core/src/firebase/firestore/model/base_path.h [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  }
+  const_iterator end() const {
+    return segments_.end();
+  }
+
+  /**
+   * Returns a new path which is the result of concatenating this path with an
+   * additional segment.
+   */
+  T Concat(const std::string& segment) const {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/base_path.h]: For Concat and DropFirst/Last, I took the general naming from [range-v3](https://ericniebler.github.io/range-v3/), which is slowly going through standardization right now.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+
+#include <algorithm>
+#include <cctype>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+#include "absl/strings/str_join.h"
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: I just ported the pieces needed here, looking forward to just having the whole Abseil. I guess I'll squash this PR into asbl_commit + everything_else_commit before merging.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  // them used in v1alpha1, but that's fine.
+
+  SegmentsT segments;
+  std::string segment;
+  segment.reserve(path.size());
+
+  // string_view doesn't have a c_str() method, because it might not be
+  // null-terminated. Assertions expect C strings, so construct std::string on
+  // the fly, so that c_str() might be called on it.
+  const auto to_string = [](const absl::string_view view) {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: This is kinda ugly. By the way, do we have any alternatives to FIREBASE_ASSERT and friends?

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    // Move operation will clear segment, but capacity will remain the same
+    // (not, strictly speaking, required by the standard, but true in practice).
+    segments.push_back(std::move(segment));
+  };
+
+  // Inside backticks, dots are treated literally.
+  bool inside_backticks = false;
+  // Whether to treat '\' literally or as an escape character.
+  bool escaped_character = false;
+  for (const char c : path) {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: I modified this loop to avoid any modification to the loop variable, and any fallthrough's inside the switch statement. This necessitated `escaped_character`. I feel it's a little easier to parse when the iteration itself is more straightforward, but feel free to push back on this.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    // std::string (and string_view) may contain embedded nulls. For full
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: Just drawing attention to this.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  // TODO(b/37244157): Make this a user-facing exception once we
+  // finalize field escaping.
+  FIREBASE_ASSERT_MESSAGE(!escaped_character,
+                          "Trailing escape characters not allowed in %s",
+                          to_string(path).c_str());
+
+  return FieldPath{std::move(segments)};
+}
+
+FieldPath FieldPath::KeyFieldPath() {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: I don't think we need a shared instance of the key field path in C++, and I don't think we need EmptyPath at all, because the default constructor does the same thing.

File: Firestore/core/src/firebase/firestore/util/firebase_assert.h [hunk 9]
```diff
@@ -75,6 +75,10 @@
     }                                                                       \
   } while (0)
 
+// Assert with custom message that is not compiled out in release builds.
+#define FIREBASE_ASSERT_MESSAGE(expression, ...) \
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/util/firebase_assert.h]: I took the liberty of defining a shorter name. If you have any suggestions on how to make it even shorter, I'll be happy to implement them, I find this SHOUTING pretty distractive.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+  std::vector<std::string> segments{"rooms", "Eros", "messages"};
+  const FieldPath path_from_segments{segments.begin(), segments.end()};
+  EXPECT_FALSE(path_from_segments.empty());
+  EXPECT_EQ(3, path_from_segments.size());
+  EXPECT_TRUE(path_from_segments.begin() + 3 == path_from_segments.end());
+
+  FieldPath copied = path_from_list;
+  EXPECT_EQ(path_from_list, copied);
+  const FieldPath moved = std::move(copied);
+  // Because FieldPath is immutable, move constructor performs a copy.
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: Just drawing attention to this.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+  EXPECT_EQ(copied, moved);
+}
+
+TEST(FieldPath, Indexing) {
+  const FieldPath path{"rooms", "Eros", "messages"};
+
+  EXPECT_EQ(path.front(), "rooms");
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: Do we really need 3 ways to say the same thing?

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+  const FieldPath path;
+  const FieldPath a{"rooms"};
+  const FieldPath ab{"rooms", "Eros"};
+  const FieldPath abc{"rooms", "Eros", "messages"};
+
+  EXPECT_EQ(a, path.Concat("rooms"));
+  EXPECT_EQ(ab, path.Concat("rooms").Concat("Eros"));
+  EXPECT_EQ(abc, path.Concat("rooms").Concat("Eros").Concat("messages"));
+  EXPECT_EQ(abc, path.Concat(FieldPath{"rooms", "Eros", "messages"}));
+  EXPECT_EQ(abc, path.Concat({"rooms", "Eros", "messages"}));
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: I think it's a pretty nice syntax.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+  ASSERT_DEATH_IF_SUPPORTED(path[0], "");
+  ASSERT_DEATH_IF_SUPPORTED(path[1], "");
+  ASSERT_DEATH_IF_SUPPORTED(path.at(0), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropFirst(), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropFirst(2), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropLast(), "");
+}
+
+TEST(FieldPath, Parsing) {
+  const auto parse = [](const std::pair<std::string, size_t> expected) {
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: Objective C uses macros for this. Helper functions are a little more verbose, but (I hope, haven't tried) mapping failure to line number should be easier.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+
+  const auto path_with_dot = FieldPath::ParseServerFormat(R"(foo\.bar)");
+  EXPECT_EQ(path_with_dot.CanonicalString(), "`foo.bar`");
+  EXPECT_EQ(path_with_dot.size(), 1);
+}
+
+// This is a special case in C++: std::string may contain embedded nulls. To
+// fully mimic behavior of Objective-C code, parsing must terminate upon
+// encountering the first null terminator in the string.
+TEST(FieldPath, ParseEmbeddedNull) {
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: Just drawing attention to this.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+  expected = make_expected("foo.bar.baz", 3);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected(R"(`.foo\\`)", 1);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected(R"(`.foo\\`.`.foo`)", 2);
+  EXPECT_EQ(expected, parse(expected));
+  expected = make_expected(R"(foo.`\``.bar)", 3);
+  EXPECT_EQ(expected, parse(expected));
+
+  const auto path_with_dot = FieldPath::ParseServerFormat(R"(foo\.bar)");
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: I added an equivalent test to Objective C. At first I thought it's a bug, but now I look at it, I guess it's reasonable.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+  str += '\0';
+  str += ".bar";
+
+  const auto path = FieldPath::ParseServerFormat(str);
+  EXPECT_EQ(path.size(), 1);
+  EXPECT_EQ(path.CanonicalString(), "foo");
+}
+
+TEST(FieldPath, ParseFailures) {
+  ASSERT_DEATH_IF_SUPPORTED(FieldPath::ParseServerFormat(""), "");
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: These don't have equivalents in Objective C.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+  EXPECT_EQ(path.DropFirst().CanonicalString(), "bar.baz");
+  EXPECT_EQ(path.DropLast().CanonicalString(), "foo.bar");
+  EXPECT_EQ(path.DropFirst().DropLast().CanonicalString(), "bar");
+  EXPECT_EQ(path.DropFirst().DropLast().CanonicalString(), "bar");
+  EXPECT_EQ(path.DropLast().DropFirst().DropLast().CanonicalString(), "");
+}
+
+TEST(FieldPath, CanonicalStringEscaping) {
+  // Should be escaped
+  EXPECT_EQ(FieldPath::ParseServerFormat("1").CanonicalString(), "`1`");
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: These don't have equivalents in Objective C.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+  EXPECT_EQ(FieldPath::ParseServerFormat("a#b").CanonicalString(), "`a#b`");
+
+  // Should not be escaped
+  EXPECT_EQ(FieldPath::ParseServerFormat("_ab").CanonicalString(), "_ab");
+  EXPECT_EQ(FieldPath::ParseServerFormat("a1").CanonicalString(), "a1");
+  EXPECT_EQ(FieldPath::ParseServerFormat("a_").CanonicalString(), "a_");
+}
+
+TEST(FieldPath, CreateKeyFieldPath) {
+  const auto key_field_path = FieldPath::KeyFieldPath();
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_path_test.cc]: These don't have equivalents in Objective C.

File: Firestore/core/test/firebase/firestore/model/resource_path_test.cc [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+#include <string>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+TEST(ResourcePath, Constructor) {
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/resource_path_test.cc]: I think there are no tests for ResourcePath in Objective C.

File: Firestore/core/test/firebase/firestore/model/resource_path_test.cc [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+  EXPECT_TRUE(empty < a);
+  EXPECT_TRUE(a < b);
+  EXPECT_TRUE(a < ab);
+
+  EXPECT_TRUE(a > empty);
+  EXPECT_TRUE(b > a);
+  EXPECT_TRUE(ab > a);
+}
+
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/resource_path_test.cc]: I don't duplicate other tests for functionality from BasePath here. Let me know if you think both FieldPath and ResourcePath should be fully tested.

File: Firestore/core/src/firebase/firestore/model/base_path.h [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+    auto concatenated = segments_;
+    concatenated.push_back(segment);
+    return T{std::move(concatenated)};
+  }
+  T Concat(std::string&& segment) const {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/base_path.h]: Just to throw an idea around, we could do operator+ instead:

```
auto path = FieldPath{"a", "b"}.Concat("c").Concat({"d", "e"});
auto path = FieldPath{"a", "b"} + "c" + Foo{"d", "e"};
```

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+  segment.reserve(path.size());
+
+  // string_view doesn't have a c_str() method, because it might not be
+  // null-terminated. Assertions expect C strings, so construct std::string on
+  // the fly, so that c_str() might be called on it.
+  const auto to_string = [](const absl::string_view view) {
+    return std::string{view.data(), view.data() + view.size()};
+  };
+  const auto finish_segment = [&segments, &segment, &path, &to_string] {
+    FIREBASE_ASSERT_MESSAGE(
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: In this and some other instances, Objective C throws, so I'm a little concerned that in C++ it's unrecoverable (OTOH, I don't know if any code in Objective C actually catches those throws).

**Comment by var-const** [comment]
Question: in Objective-C, the fact that FIRFieldPath uses FSTFieldPath inside is completely invisible in the header file. @wilhuff, do we want to imitate that with pimpl?

File: Firestore/core/src/firebase/firestore/model/base_path.h (hunk 21 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/base_path.h]: I'm not completely against operator overloading but I think if there's a reasonable way to represent something with method calls (and maintain parity with the other ports) we should do that instead.

File: Firestore/core/src/firebase/firestore/model/base_path.h (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/base_path.h]: My first choices for this would be `append` or `child` from the other ports. Indeed I'd take this as an opportunity to line the ports up more closely and make them all standardize on one name.

File: Firestore/core/src/firebase/firestore/model/base_path.h [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+    auto concatenated = segments_;
+    concatenated.insert(concatenated.end(), path.begin(), path.end());
+    return T{std::move(concatenated)};
+  }
+
+  /**
+   * Returns a new path which is the result of dropping the first n segments of
+   * this path.
+   */
+  T DropFirst(const size_t n = 1) const {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/base_path.h]: This should be PopFirst to match the other ports. In general the first inspiration for names should be the other ports.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+
+/**
+ * True if the string could be used as a segment in a field path without
+ * escaping.
+ */
+bool IsValidIdentifier(const std::string& segment) {
+  if (segment.empty()) {
+    return false;
+  }
+  if (segment.front() != '_' && !std::isalpha(segment.front())) {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: `std::isalpha` is locale sensitive and therefore incorrect.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+    return false;
+  }
+  if (std::any_of(segment.begin(), segment.end(), [](const unsigned char c) {
+        return c != '_' && !std::isalnum(c);
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: `std::isalnum` is also locale sensitive and incorrect.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+ * escaping.
+ */
+bool IsValidIdentifier(const std::string& segment) {
+  if (segment.empty()) {
+    return false;
+  }
+  if (segment.front() != '_' && !std::isalpha(segment.front())) {
+    return false;
+  }
+  if (std::any_of(segment.begin(), segment.end(), [](const unsigned char c) {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: While this functional style might be a local improvement, it hurts cross-platform comprehension. The simple for loop is easier to understand and port.

File: Firestore/core/src/firebase/firestore/util/firebase_assert.h (hunk 9 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/firebase_assert.h]: This change is in one of the other PRs too.

Unfortunately, these SHOUTING names that are already in the Firebase C++ SDK so we're stuck with them.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc (hunk 10 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/model/field_path_test.cc]: I'm torn about this. Let's discuss.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc (hunk 11 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/model/field_path_test.cc]: No. I suggest removing `at`.

`front` and `[0]` are necessary because the analogous `back` and `path[path.size() - 1]` are not so easy.

File: Firestore/core/test/firebase/firestore/model/resource_path_test.cc (hunk 20 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/model/resource_path_test.cc]: I think it's reasonable to test the parts that are different and rely on the shared implementation.

File: Firestore/third_party/abseil-cpp/absl/numeric/CMakeLists.txt [hunk 27]
```diff
@@ -0,0 +1,62 @@
+#
```
Comment (wilhuff) [inline Firestore/third_party/abseil-cpp/absl/numeric/CMakeLists.txt]: This should be done in a separate CL, based of the `abseil-mirror` branch and then merged to `master`. This way we have pristine sources and we can track how we got them. Also we should be importing from the same version.

Comment (var-const) [reply Firestore/third_party/abseil-cpp/absl/numeric/CMakeLists.txt]: > Also we should be importing from the same version.

How can I find out the exact Abseil version to use?

**Comment by wilhuff** [comment]
Regarding this question:
> Question: in Objective-C, the fact that FIRFieldPath uses FSTFieldPath inside is completely invisible in the header file. @wilhuff, do we want to imitate that with pimpl?

We still need to get our story straight re how we're going to ship public headers. Ideally yes, we would hide this. I wouldn't tackle FIRFieldPath in this PR anyway though.

File: Firestore/core/src/firebase/firestore/model/base_path.h (hunk 2 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/base_path.h]: Thanks for explaining the rationale:

* removed `at`;
* changed `front`/`back` into `first/last_segment` (snake case because arguably they are properties);
* changed `Concat` to `Append` in the style of Android (I prefer it to `Child`).

File: Firestore/core/src/firebase/firestore/model/base_path.h (hunk 21 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/base_path.h]: It also has the downside in that binary operators in C++ more often return references, not new instances.

File: Firestore/core/src/firebase/firestore/model/base_path.h (hunk 23 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/base_path.h]: Done.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 24 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: Thanks for catching that.

Note the added comment. Strictly speaking, in C++ it's not guaranteed that 'a'-'z' and 'A'-'Z' are contiguous ranges. However, the only system I'm aware of where this expectation is broken is pretty exotic (EBCDIC used on old IBM mainframes), so I think it makes sense to ignore that.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 26 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: Rolled back.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 25 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: Done.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 22 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: Thanks for the explanation.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 6 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: Rolled back.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 8 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: Done.

File: Firestore/core/src/firebase/firestore/util/firebase_assert.h (hunk 9 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/util/firebase_assert.h]: > This change is in one of the other PRs too.

Can you link that PR? Did you mean to imply that I should roll this back?

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc (hunk 10 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/model/field_path_test.cc]: For the record: during offline discussion, it was decided to make Field/ResourcePath reassignable. This leaves it up to the code using paths to ensure they are used in a thread-safe way.

File: Firestore/core/test/firebase/firestore/model/field_path_test.cc (hunk 11 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/model/field_path_test.cc]: Removed `at`.

File: Firestore/third_party/abseil-cpp/absl/numeric/CMakeLists.txt (hunk 27 above)
Comment (var-const) [reply Firestore/third_party/abseil-cpp/absl/numeric/CMakeLists.txt]: This no longer applies. Gil, thanks for helping with Abseil.

File: Firestore/core/src/firebase/firestore/util/firebase_assert.h (hunk 9 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/firebase_assert.h]: It's here https://github.com/firebase/firebase-ios-sdk/pull/733/files#diff-398e46f358d36158bce3e568e0a6ae48

You don't need to roll back, just be careful to merge correctly (and line up on the names now).

**Comment by wilhuff** [comment]
Please run `./scripts/style.sh` (that's what travis is complaining about).
